### ISSUE TITLE
Minor hashbang-related inconsistency between the analyzer & spec and the compiler.

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/sdk#52331]
Consider the following two files:

```dart
// This is a single file called 'foo.dart'.
part 'foo.gen.dart';

void main() {
  run();
}
```

and

```dart
#! foo
// This is a single file called 'foo.gen.dart'.
part of 'foo.dart';

void run() {
  print("foofoo");
}
```

The analyzer emits an error on the keyword `part` in `foo.gen.dart` that says:

```
Directives must appear before any declarations.

Try moving the directive before any declarations.
```

This seems to suggest that `foo.dart` should not compile successfully. However, we can run `foo.dart` using `dart foo.dart` and it terminates successfully and prints `foofoo`.

According to the spec, the analyzer exhibits correct behavior and `foo.dart` should not compile because the grammar does not imply that a partHeader with a `SCRIPT_TAG` is in the Dart language.

---

I'm not too familiar with how the shell treats hashbangs and I might be missing some details, but, in my opinion, it looks like the way the compiler behaves allows for a cleaner separation of concerns and It might make sense to keep its behavior and adjust the analyzer and the spec:

```bnf
utfSublanguage ::= FEFF? shellSublanguage
shellSublanguage ::= SCRIPT_TAG? dartSublanguage
dartSublanguage ::= ...
...
```

because it would allow us to factor out the shell-related parts out of the core dart language grammar instead of only allowing `libraryDefinition`s to be members of the `"shellSublanguage"`.

**Comment by lrhn** [comment]
Yep, I agree that the analyzer is correct. A part file cannot start with `#!`.

Most likely the compilers never check, they just treat all initial `#!` lines as comments, and ignore them completely.

I don't see any particular value in making it an error, so maybe we should allow part files to start with a script tag too.
@eernstg

**Comment by eernstg** [comment]
> I don't see any particular value in making it an error, so maybe we should allow part files to start with a script tag too.

We could do that.

Just testing, it is actually possible to run a program which is determined by a part file: The part file is `part of` some library; that library is determined to the same extent as any other Dart source file if the part uses a URI to denote it; and `dart myPart.dart` _will_ run the program when `myPart.dart` specifies a library which can be the entry point for a Dart program execution (that is, it exports a top-level function named `main`, cf. [this spec paragraph](https://github.com/dart-lang/language/blob/c16cf29306145229ecfaabea8160a0564f92c883/specification/dartLangSpec.tex#L20082)).

In other words, putting a script tag into a part makes sense for the exact same reason that it makes sense to put it into a library, and it could even be used to run the same dart program in several different ways, based on different script tags in different parts. ;-)

**Comment by lrhn** [comment]
Being able to use a part file as entry point sounds like being outside of the specified behavior. Unless we've been very vague in specifying, which never happens!

Even then, the `#!` shell trigger can be on any file, it doesn't have to start running the specified executable with that file, it's just easier to do that.

So we could allow a part file to start with a script tag.
Or we could make the front end err if it sees a part file with a script tag.
Or we can close our eyes and keep pretending nobody ever noticed.

In the end, I didn't it'll ever matter.

**Comment by johnniwinther** [comment]
The ability to run a program from a part is accidental and only works if the `part of` declaration refers to the URI and not just the library name. This is an inadvertent consequence of supporting seeing part files before their containing library during incremental compilation. I'd very much like _not_ to have to support this continuously since it is part of a long standing technical debt within the CFE.

**Comment by eernstg** [comment]
@johnniwinther, considering that the analyzer rejects a part with a script tag, it should be safe to make it an error with the CFE as well.

In that case I'd suggest that we put the label 'implementation' on this issue and use it to be the issue that requests an implementation of this error in the CFE at some point in time where it fits the priorities. WDYT?

**Comment by johnniwinther** [comment]
SGTM

**Comment by eernstg** [comment]
Done in https://dart-review.googlesource.com/c/sdk/+/315420.
